gzip.go

```diff
@@ -63,35 +69,93 @@ func addLevelPool(level int) {
 // GzipResponseWriter provides an http.ResponseWriter interface, which gzips
 // bytes before writing them to the underlying response. This doesn't close the
 // writers, so don't forget to do that.
+// It can be configured to skip response smaller than minSize.
 type GzipResponseWriter struct {
 	http.ResponseWriter
 	index int // Index for gzipWriterPools.
 	gw    *gzip.Writer
+
+	code int // Saves the WriteHeader value.
+
+	minSize      int    // Specifed the minimum response size to gzip. Only the first Write call is checked.
+	buff         []byte // Holds the first part of the write before the reaching the minSize or the end of the write.
+	bytesWritten int    // Keep trace of the numbers of bytes written.
 }
 
 // Write appends data to the gzip writer.
 func (w *GzipResponseWriter) Write(b []byte) (int, error) {
```

Comment:
I think we need to reconsider this function. It's quite busy and stateful. It would be better to split it into two parts:

This func is:
1. If gw != nil (i.e. we are gzipping), write and return
2. Append to buf
3. If len(buf) has exceeded minGzipBytes, call startGzip

Where `startGzip` is:
1. Write headers
2. Set gw
3. Flush buf

Am I missing anything here?